### PR TITLE
Add support for embedding domains in manifests

### PR DESCRIPTION
**Pull request by nbougalis**

### CODE REVIEW

File: src/ValidatorKeys.cpp [hunk 1]
```diff
@@ -214,20 +259,58 @@ ValidatorKeys::revoke ()
...
+        // thorough regex can be used or this check can just be removed.
+        static boost::regex const re(
+                "^"                     // Beginning of line
+                "("                     // Hostname or domain name
+                "(?!-)"                 //  - must not begin with '-'
+                "[a-zA-Z0-9-]{1,63}"    //  - only alphanumeric and '-'
+                "(?<!-)"                //  - must not end with '-'
+                "\\."                   // segment separator
+                ")+"                    // 1 or more segments
+                "[A-Za-z]{2,63}"        // TLD
```
Comment (mellery451) [inline src/ValidatorKeys.cpp]: hmm - TLDs can be punycode, right? Do we care about ascii-only for this field? We're not going to try to resolve it, are we? Alternatively, we might consider just IDNA2008 encoding anything high-bit. I added a testcase that seems like it should work, but does not currently:  https://github.com/mellery451/validator-keys-tool/commit/e1917fc3a85cd0582fc4ef5796e5fc0fd0b6352b

File: src/ValidatorKeysTool.cpp [hunk 2]
```diff
@@ -152,6 +155,82 @@ void createRevocation (boost::filesystem::path const& keyFile)
...
+    else
+    {
+        std::cout << "The domain name has been set.\n";
+        std::cout << "You should publish a domain attestation token to allow ";
+        std::cout << "others to associate this validator with your domain.";
+        attestDomain(keys);
+    }
+
+    // Update key file with new token sequence
+    keys.writeToFile (keyFile);
```
Comment (ximinez) [inline src/ValidatorKeysTool.cpp]: Suggestion: Write the file before printing the output, so the data won't be lost in the extremely unlikely case that the app crashes during the output.

File: src/ValidatorKeysTool.cpp [hunk 3]
```diff
@@ -170,6 +249,40 @@ void signData (std::string const& data,
...
+
+    auto keys = ValidatorKeys::make_ValidatorKeys (keyFile);
+
+    auto const m = keys.manifest();
+
+    if (m.empty())
+    {
+        std::cout << "The last manifest generated is unavailable. You can\n";
+        std::cout << "generate a new one.\n\n";
+        return;
```
Comment (ximinez) [inline src/ValidatorKeysTool.cpp]: Suggestion: Provide the command line to generate a new manifest. eg. `generate a new one with the "create_token" command.`

File: src/ValidatorKeys.h [hunk 4]
```diff
@@ -130,6 +136,28 @@ class ValidatorKeys
     {
         return revoked_;
     }
+
+    /** Returns the domain associated with this key, if any */
+    std::string
```
Comment (mellery451) [inline src/ValidatorKeys.h]: I think this could return `const&`

File: src/ValidatorKeysTool.cpp [hunk 5]
```diff
@@ -223,10 +361,14 @@ void printHelp (const boost::program_options::options_description& desc)
...
         << "Commands: \n"
-           "     create_keys        Generate validator keys.\n"
-           "     create_token       Generate validator token.\n"
-           "     revoke_keys        Revoke validator keys.\n"
-           "     sign <data>        Sign string with validator key.\n";
+           "     create_keys         Generate validator keys.\n"
+           "     create_token        Generate validator token.\n"
+           "     revoke_keys         Revoke validator keys.\n"
+           "     sign <data>         Sign string with validator key.\n"
+           "     show_manifest       Displays the last generated manifest\n"
```
Comment (mellery451) [inline src/ValidatorKeysTool.cpp]: according to `commandArgs`, `show_manifest` takes an argument - type perhaps?

**Pull request merged**